File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.cpp
class Solution {
public:
    vector<int> validSequence(string word1, string word2) {
        int n = word1.size();
        int m = word2.size();
        
        vector<int> last(m, -1);
        
        int ptr = n - 1;
        for (int j = m - 1; j >= 0; --j) {
            while (ptr >= 0 && word1[ptr] != word2[j]) {
                ptr--;
            }
            if (ptr >= 0) {
                last[j] = ptr;
                ptr--;
            }
        }

        vector<int> result;
        bool changed = false;
        int word1_idx = 0;

        for (int word2_idx = 0; word2_idx < m; ++word2_idx) {
            bool found = false;

            while (word1_idx < n) {
                if (word1[word1_idx] == word2[word2_idx]) {
                    result.push_back(word1_idx);
                    word1_idx++;
                    found = true;
                    break;
                }

                if (!changed) {

                    if (word2_idx == m - 1 || (last[word2_idx + 1] > word1_idx)) {
                        result.push_back(word1_idx);
                        word1_idx++;
                        changed = true;
                        found = true;
                        break;
                    }
                }

                word1_idx++;
            }

            if (!found) {
                return {};
            }
        }

        return result;
    }
};

File: 3302-find-the-lexicographically-smallest-valid-sequence/3302-find-the-lexicographically-smallest-valid-sequence.py
class Solution:
    def validSequence(self, word1: str, word2: str) -> list[int]:
        n, m = len(word1), len(word2)
        last = [-1] * m
        
        ptr = n - 1
        for j in range(m - 1, -1, -1):
            while ptr >= 0 and word1[ptr] != word2[j]:
                ptr -= 1
            if ptr >= 0:
                last[j] = ptr
                ptr -= 1

        result = []
        changed = False
        word1_idx = 0

        for word2_idx in range(m):
            found = False

            while word1_idx < n:
                if word1[word1_idx] == word2[word2_idx]:
                    result.append(word1_idx)
                    word1_idx += 1
                    found = True
                    break
                
                if not changed:
                    if word2_idx == m - 1 or last[word2_idx + 1] > word1_idx:
                        result.append(word1_idx)
                        word1_idx += 1
                        changed = True
                        found = True
                        break

                word1_idx += 1

            if not found:
                return []

        return result